OSDOCS-18911 OLMv1 ClusterObjectSets (#112334)
openshift/openshift-docs: modules/olmv1-clusterobjectsets.adoc

// Module is included in the following assemblies:
//
// * openshift-docs/exxtensions/arch/operator-controller.adoc

:_mod-docs-content-type: CONCEPT
[id="olmv1-clusterobjectsets_{context}"]
= ClusterObjectSets

:FeatureName: ClusterObjectSets

> include::snippets/technology-preview.adoc[]

Reviewer: I think we need speak here more about what ClusterObjectSets represents. The way that its written seems like it is like the CE and it is not. So we need to clarify

IMO, I think of each ClusterObjectSet as an immutable snapshot of “what this extension version should look like on the cluster.”

We have https://github.com/operator-framework/operator-controller/blob/main/docs/draft/concepts/clusterobjectsets.md it might be helpful. But for end users I think we do not go in all details, we might need only calrify that it is internal, imutable, why it exist and how  OLM uses it in overall.

A lifecycle state:: Either `Active` (being reconciled) or `Archived` (inactive). The  lifecycle state transitions from `Active` to `Archived` but not back.

A list of phases:: Objects within a ClusterObjectSet are organized into phases. Each phase groups related resources, and phases are applied sequentially. Within a phase, all objects are applied simultaneously in no particular order.

A collision protection strategy:: Controls whether a ClusterObjectSet can adopt pre-existing objects on the cluster. 

Status conditions:: Indicates whether the revision is actively rolling out and reports rollout progress, availability, and success.

Additionally, support for management of large object bundles created by ClusterObjectSet is provided. ClusterObjectSets embed Kubernetes manifests in `.spec.phases[].objects[].object`. With up to 20 phases and 50 objects per phase, they can exceed the etcd 1.5 MiB object size limit. Large Operators with many CRDs, Deployments, RBAC rules, and webhook configurations frequently hit this limit, causing the API server to reject the ClusterObjectSet and fail installation or upgrade.           

>                        

Reviewer: ```suggestion
As part of  {olmv1-first}, when you install or upgrade content with a `ClusterExtension`, the operator-controller creates `ClusterObjectSet` resources for you.
You only manage `ClusterExtension`. Each `ClusterObjectSet` is one install or upgrade step: the manifests for that version, applied in a safe order.

A `ClusterObjectSet` is a cluster-scoped resource. It is not content you install. It is the controller’s internal record for one revision of your `ClusterExtension`.

The controller groups manifests into _phases_ (for example, CRDs before Deployments) and applies each phase in order. It waits until a phase is ready before starting the next. That avoids half-finished installs.

When you install content, the controller creates the first `ClusterObjectSet` (revision `1`). When the bundle version changes, it creates a new revision. During an upgrade, the old and new revision can both be active until the new one finishes; then older ones are archived (up to five kept for history). Check progress on the `ClusterExtension` in `status.activeRevisions`, or run `oc get clusterobjectsets -l olm.operatorframework.io/owner-name=<extension_name>`.

Each `ClusterObjectSet` has a fixed `revision` number and `phases` that do not change after creation. `lifecycleState` is `Active` while rolling out, then `Archived`. `status` conditions such as `Progressing`, `Available`, and `Succeeded` show whether that step succeeded. At most one `ClusterObjectSet` manages a given cluster object at a time. Very large bundles store manifests in `Secret` objects so size limits do not block install or upgrade.

[NOTE]
====
Archived `ClusterObjectSet` revisions stay on the cluster as a history of what was applied for that `ClusterExtension`.
====
```

Reviewer: Maybe we could simplify and keep it less verbose with only the most important info for consumers? It is an internal resource so we might not need describe all details. WDYT @perdasilva ?

// Module is included in the following assemblies:
//

> // * openshift-docs/exxtensions/arch/operator-controller.adoc

Reviewer 2: Nit: typo — `exxtensions` should be `extensions`.

:_mod-docs-content-type: CONCEPT
[id="olmv1-clusterobjectsets_{context}"]
= ClusterObjectSets

:FeatureName: ClusterObjectSets
include::snippets/technology-preview.adoc[]

[role="_abstract"]

> As part of {olmv1-first}, the operator-controller creates `ClusterObjectSet` resources when you install or upgrade content with a `ClusterExtension`. `ClusterObjectSet` enables safe, phased rollouts of Kubernetes resources. 

Reviewer 3: Needs nouns.

```suggestion
As part of {olmv1-first}, the operator-controller creates `ClusterObjectSet` resources when you install or upgrade content with a `ClusterExtension` object. `ClusterObjectSet` objects enable safe, phased rollouts of Kubernetes resources. 
```

> As part of  {olmv1-first}, the operator-controller creates `ClusterObjectSet` resources when you install or upgrade content with a `ClusterExtension`. `ClusterObjectSet` enables safe, phased rollouts of Kubernetes resources. 

Reviewer 2: Trailing double space before `{olmv1-first}` on this line.

> A `ClusterObjectSet` is a cluster-scoped resource. It is not content that you install. It is the controller’s internal record for one revision of your `ClusterExtension`. It rolls out related resources sequentially with built-in readiness checks between phases. 

Reviewer 3: Needs nouns.

```suggestion
A `ClusterObjectSet` object is a cluster-scoped resource. It is not content that you install. It is the controller’s internal record for one revision of your `ClusterExtension` object. It rolls out related resources sequentially with built-in readiness checks between phases. 
```

> A `ClusterObjectSet` solves several problems when managing sets of related Kubernetes resources: 

Reviewer 3: ```suggestion
A `ClusterObjectSet` object solves several problems when managing sets of related Kubernetes resources: 
```

Immutable revision content:: The revision number, phases, objects, and collision protection strategy are immutable, that is, they cannot change once set. This provides a clear audit trail of what was deployed at each revision.  
                                                      
Phased rollout:: Resources are grouped into phases, for example, CRDs before Deployments, and are deployed sequentially. A phase only progresses after all its objects pass readiness probes. Manifests are organized into phases, applied to the cluster, and must pass some status probes before they progress to the next phase. This avoids incomplete installs.

> 

Reviewer 2: The code also supports user-defined custom readiness probes via `progressionProbes` (using `ConditionEqual`, `FieldsEqual`, or `FieldValue` assertions). Saying only "built-in readiness checks" undersells the feature — operators with custom resources can define their own probes to gate phase progression.

Separately, `progressDeadlineMinutes` (10-720 min) lets users set a rollout timeout. Both are experimental, but worth at least a mention since the doc already covers a Tech Preview feature.

See `ClusterObjectSetSpec.ProgressionProbes` and `ClusterObjectSetSpec.ProgressDeadlineMinutes` in `api/v1/clusterobjectset_types.go`.

Safe transitions:: During upgrades, both old and new revisions remain active until the new revision succeeds. Object ownership transitions between revisions automatically.          

Single ownership:: Each resource can only be managed by one `ClusterObjectSet` at a time, preventing conflicts.                                                                

> Large resource support:: Object manifests can be stored inline or externalized into Secrets, allowing bundles larger than the etcd 1.5 MiB object size limit. 

Reviewer 3: ```suggestion
Large resource support:: Object manifests can be stored inline or externalized into secrets, allowing bundles larger than the etcd 1.5 MiB object size limit. 
```

[NOTE]
====

> Archived `ClusterObjectSet` revisions stay on the cluster as a history of what was applied for that `ClusterExtension`. 

Reviewer 3: ```suggestion
Archived `ClusterObjectSet` revisions stay on the cluster as a history of what was applied for that `ClusterExtension` object. 
```

Reviewer 3: I will let you address the rest of these as appropriate.

> Archived `ClusterObjectSet` revisions stay on the cluster as a history of what was applied for that `ClusterExtension`. This provides an historical footprint of the changes applied to the cluster as part of the Operator. 

Reviewer: ```suggestion
Archived `ClusterObjectSet` revisions stay on the cluster as a history of what was applied for that `ClusterExtension`. 
```

I think the rest is not required ( we are just saying the same )
Also, we should avoid use Operator and isntead we should use like Content

Not all content managed by OLM is necessary and Operator. 
More over OLM has the goal in the roadmap of manage HelmChart

Reviewer 4: > Not all content managed by OLM is necessary and Operator.
> More over OLM has the goal in the roadmap of manage HelmChartThis is forward looking

I think this is forward looking and until its in place, using future verbiage could be a source of confusion. It should be ok to use operators for now since that's the only thing that can be installed and its probably closer to the user's expectation/current mental model. Of course, once those features make it in, we'd need to update the docs. But, I'll leave it to docs to decide what makes more sense. Though, we should take care not leak roadmap into the docs (we may set expectations we might never meet)...

====

Each `ClusterObjectSet` contains the following content:

A revision number:: A permanent, sequential integer that identifies the version. When the bundle version changes, it creates a new revision. During an upgrade, the old and new revision can both be active until the new one finishes. Older revisions are archived.

A lifecycle state:: This shows the status of the `ClusterObjectSet`, which can be `Active` while rolling out, then `Archived`. The  `lifecycleState` transitions from `Active` to `Archived` but not back.

> A list of phases:: Objects within a `ClusterObjectSet` are organized into phases. Each phase groups related resources, and phases are applied sequentially. Within a phase, all objects are applied simultaneously in no particular order.

Reviewer 2: "Active while rolling out" is slightly misleading. Per the API, `Active` means the revision is "actively managed and reconciled" — it stays Active from creation through successful rollout and beyond, until explicitly archived after a successor completes. It is not limited to the rolling-out window.

Suggestion: _"which can be `Active` (managed and reconciled) or `Archived`"_

Author: Made this change.

A collision protection strategy:: Controls whether a `ClusterObjectSet` can adopt pre-existing objects on the cluster. 

Status conditions:: Indicates whether the revision is actively rolling out and reports rollout progress, availability, and success.

At most, one `ClusterObjectSet` manages a given cluster object at a time. Very large bundles store manifests in `Secret` objects so that size limits do not block installs or upgrades.

> Additionally, support for management of large object bundles created by `ClusterObjectSet` is provided. `ClusterObjectSet` embeds Kubernetes manifests in `.spec.phases[].objects[].object`. With up to 20 phases and 50 objects per phase, they can exceed the etcd 1.5 MiB object size limit. Large Operators with many CRDs, Deployments, RBAC rules, and webhook configurations frequently hit this limit, causing the API server to reject the `ClusterObjectSet` and fail installation or upgrade.           

Reviewer: Let not add things like

```
With up to 20 phases and 50 objects per phase
```

The COs are for internal work and it can easily change.

Reviewer: Also, lets not use the term Operators.

Reviewer 4: Maybe we could just say "Currently with up to 20 phases...". While COS is used as an implementation/internal detail of how OLM does what it does, the API is still present on the cluster and can be used independently - any changes applied that come will be non-breaking (i.e. the API will evolve in the same way any other API that we ship does)

>                        

Reviewer 2: "per-object content references" is vague — earlier in the doc you correctly say "Secrets". The actual mechanism is an `ObjectSourceRef` pointing to a specific key within a named `Secret`. Consider keeping the terminology consistent, e.g.: _"objects can be externalized into `Secret` resources using per-object references"_.
